File: java/src/main/java/practice/concurrency/lrucache/LRUCache.java
package practice.concurrency.lrucache;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

public class LRUCache<K, V> {
    private final int capacity;
    
    private final ReentrantLock lock = new ReentrantLock();
    private final Map<K, Node<K, V>> cache = new HashMap<>();
    private final Node<K, V> head = new Node<>(null, null);
    private final Node<K, V> tail = new Node<>(null, null);

    public LRUCache(int capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException("Capacity must be larger than 0");
        }
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }

    public V get(K key) {
        lock.lock();
        try {
            Node<K, V> node = cache.get(key);
            if (node == null) {
                return null;
            }

            moveToFront(node);
            return node.value;
        } finally {
            lock.unlock();
        }
    }

    public void put(K key, V value) {
        lock.lock();
        try {
            Node<K, V> node = cache.get(key);
            if (node != null) {
                node.value = value;
                moveToFront(node);
                return;
            }

            Node<K, V> newNode = new Node<>(key, value);
            cache.put(key, newNode);
            addToFront(newNode);

            if (cache.size() > capacity) {
                Node<K, V> oldest = removeLast();
                cache.remove(oldest.key);
            }
        } finally {
            lock.unlock();
        }
    }

    private void moveToFront(Node<K, V> node) {
        remove(node);
        addToFront(node);
    }

    private void addToFront(Node<K, V> node) {
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }

    private void remove(Node<K, V> node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private Node<K, V> removeLast() {
        Node<K, V> last = tail.prev;
        remove(last);
        return last;
    }

    private static class Node<K, V> {
        K key;
        V value;
        Node<K, V> prev;
        Node<K, V> next;

        Node(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }
}

File: go/exercises/lrucache/lru_cache.go
package lrucache

import "sync"

type Node[K comparable, V any] struct {
	key   K
	value V
	prev  *Node[K, V]
	next  *Node[K, V]
}

type LRUCache[K comparable, V any] struct {
	capacity int
	cache    map[K]*Node[K, V]
	head     *Node[K, V]
	tail     *Node[K, V]
	mu       sync.Mutex
}

func NewLRUCache[K comparable, V any](capacity int) *LRUCache[K, V] {
	if capacity <= 0 {
		panic("Capacity must be larger than 0")
	}

	head := &Node[K, V]{}
	tail := &Node[K, V]{}

	head.next = tail
	tail.prev = head

	return &LRUCache[K, V]{
		capacity: capacity,
		cache:    make(map[K]*Node[K, V]),
		head:     head,
		tail:     tail,
	}
}

func (l *LRUCache[K, V]) Get(key K) (V, bool) {
	l.mu.Lock()
	defer l.mu.Unlock()

	node, found := l.cache[key]

	if !found {
		var zero V
		return zero, false
	}

	l.moveToFront(node)
	return node.value, true
}

func (l *LRUCache[K, V]) Put(key K, value V) {
	l.mu.Lock()
	defer l.mu.Unlock()

	node, found := l.cache[key]

	if found {
		node.value = value
		l.moveToFront(node)
		return
	}

	newNode := &Node[K, V]{
		key:   key,
		value: value,
	}

	l.cache[key] = newNode
	l.moveToFront(newNode)

	if len(l.cache) > l.capacity {
		oldest := l.removeLast()
		delete(l.cache, oldest.key)
	}
}

func (l *LRUCache[K, V]) moveToFront(node *Node[K, V]) {
	l.remove(node)
	l.addToFront(node)
}

func (l *LRUCache[K, V]) addToFront(node *Node[K, V]) {
	node.next = l.head.next
	node.prev = l.head

	l.head.next.prev = node
	l.head.next = node
}

func (l *LRUCache[K, V]) remove(node *Node[K, V]) {
	node.prev.next = node.next
	node.next.prev = node.prev
}

func (l *LRUCache[K, V]) removeLast() *Node[K, V] {
	node := l.tail.prev
	l.remove(node)
	return node
}
